Better Cython integration

arthurp opened the pull request.

arthurp committed:
Better Cython integration.

pyrito commented on libkatana_python_native/include/katana/python/CythonIntegration.h:
```diff
@@ -8,12 +8,27 @@
+///   ownership.
+/// - a static method _make_from_address_shared which takes shared_ptr<T>*
+///   as an int, and returns a new Python object wrapping it. Creates a new
+///   shared_ptr internally.
```
What happens to the ref count on the original shared_ptr that is passed in? Does that counter get incremented? What is the internal shared_ptr needed for?

arthurp replied:
`shared_ptr` represents the concept of a reference counted pointer. If Cython holds a reference it will (if written even vaguely correctly) hold a reference count as well.

arthurp replied:
I'll remove the comment about the internal shared pointer. That's an implementation detail.

pyrito commented on libkatana_python_native/include/katana/python/CythonIntegration.h:
```diff
@@ -8,12 +8,27 @@
+///   shared_ptr internally.
+/// - a static method _make_from_address_unique which takes unique_ptr<T>*
+///   as an int, and returns a new Python object wrapping it. Moves out of
+///   the unique_ptr.
```
Wdym by "moves out of the unique_ptr"? Does this mean we don't take ownership, rather make a copy?

arthurp replied:
I mean it takes ownership of the referenced object via `std::move`. Bad wording sorry.

arthurp committed:
Actually check for the environment variable used.

arthurp committed:
Move from_graphml to pybind11. Fix ErrorHandling ref counting.

ErrorHandling was returning a py::object as a py::handle. This
returns the handle without an associated reference count and
results in an early free when the py::object goes out of scope.

pyrito commented on libkatana_python_native/include/katana/python/CythonIntegration.h:
```diff
@@ -119,6 +134,7 @@ namespace katana {
 // Add Cython classes here as needed. Remove them when they are moved to
 // pybind11.
 CYTHON_REFERENCE_SUPPORT(katana::PropertyGraph, "katana.local", "Graph");
+CYTHON_REFERENCE_SUPPORT(katana::TxnContext, "katana.local", "TxnContext");
```
Can I use this for the distributed case or would that need to be wrapped again?

arthurp replied:
You can use this.

pyrito commented on python/katana/local/_graph.pyx:
```diff
@@ -592,3 +602,12 @@ cdef class Graph(GraphBase):
         raise NotImplementedError()
+
+
+cdef class TxnContext:
+    @property
+    def __katana_address__(self):
+        """
+        Internal.
+        """
+        return <uint64_t>&self._txn_ctx
```
Why was this needed?

arthurp replied:
To enable the cython integration above.

pyrito replied:
Is this all that's needed?

arthurp replied:
For `CythonReference` yes. If you need `MakeCythonWrapper` you also need to implement at least one of the `_make_from_address` methods.

pyrito reviewed (approved):
LGTM - everything looks sane to me
